environment.yaml

```diff
@@ -0,0 +1,35 @@
+# reasons you might want to use `environment.yaml` instead of `requirements.txt`:
+# - pip installs packages in a loop, without ensuring dependencies across all packages
+#   are fulfilled simultaneously, but conda achieves proper dependency control across
+#   all packages
+# - conda allows for installing packages without requiring certain compilers or
+#   libraries to be available in the system, since it installs precompiled binaries
+
+name: myenv
+
+channels:
+  - pytorch
+  - conda-forge
+  - defaults
+
+dependencies:
+  - pytorch>=1.10.0
+  - torchvision>=0.11.0
+  - pytorch-lightning=1.9
+  - torchmetrics=0.11
+  - hydra-core=1.3
+  - rich=13.3
+  - pre-commit=3.0
+  - pytest=7.2
+
+  # --------- loggers --------- #
+  # - wandb
+  # - neptune-client
+  # - mlflow
+  # - comet-ml
+
+  - pip>=23.0
+  - pip:
+      - hydra-optuna-sweeper==1.2.0
+      - hydra-colorlog==1.2.0
+      - pyrootutils==1.0.4
```

Comment:
I think I would prefer to remove the hardcoding of versions of all the dependencies since it's bothersome to regularly update them in the template. How about just encouraging to specify versions trough comment?

Reply:
I think that's fine for the pip packages, since they're hardcoded to a single version and can't be weaved into the dependency graph solving that conda does, but for the conda packages I would suggest a different approach than your suggestion. I hear you and sympathize with it being bothersome to regularly update versions, but I think there's a middle ground here. First off, I would highly advise against removing any sort of version specification as it would defeat the purpose of having a reproducible environment so that others can have the same experience with the template. Also, it may result in a situation where all version-unspecified packages install their latest major versions which can break things in the template. This is why I didn't hardcode the conda packages to a single version (`package==x.x.x`) yet still used `package=x.x` or `package>=x.x`. This allows for flexibility in allowing for new minor versions of packages to be installed while guaranteeing all packages to be at some minimally compatible version for the template to work as expected. [This website has a pretty good explanation on it](https://carpentries-incubator.github.io/introduction-to-conda-for-data-scientists/04-sharing-environments/index.html#always-version-control-your-environmentyml-files) and [this website describes how conda does package semantic version matching](https://docs.conda.io/projects/conda-build/en/latest/resources/package-spec.html#package-match-specifications). I think we can ease the version specification further without making it completely version-unspecified. For example, we can probably have:

```yaml
  - pytorch>=1.10.0
  - torchvision>=0.11.0
  - pytorch-lightning=1.*
  - torchmetrics=0.*
  - hydra-core=1.*
  - rich=13.*
  - pre-commit=3.*
  - pytest=7.*
  - pip>=23.0
  - pip:
      - hydra-optuna-sweeper
      - hydra-colorlog
      - pyrootutils
```

or equivalently (or at least I think it is):
```yaml
  - pytorch>=1.10
  - torchvision>=0.11
  - pytorch-lightning=1
  - torchmetrics=0
  - hydra-core=1
  - rich=13
  - pre-commit=3
  - pytest=7
  - pip>=23
  - pip:
      - hydra-optuna-sweeper
      - hydra-colorlog
      - pyrootutils
```

This keeps the dependencies in the same major versions across all users, but allows for flexibility for different minor and patch versions of packages where backwards compatibility is guaranteed for newer minor and/or patch versions. How does this sound?

Reply:
I like this. I will add some explanation comment for this approach later.

Reply:
Totally forgot I hadn't made the changes above yet! I just committed the changes now :) Let me know what you think!